### PR TITLE
[2.19.x] DDF-6618 / G-9917 Adds proper namespace support to CSW GetRecordsMess…

### PR DESCRIPTION
**Pull request by jlcsmith**
…ageBodyReader

#### What does this PR do?
Fixes a bug with the CSW Source in the case that the CSW server sends back a GetRecordsResponse with a non-"csw" namespace prefix.  Now the MessageBodyReader properly handles namespace prefix mappings.

#### Who is reviewing it? 
@jrnorth @millerw8 @maiiiiiiii 

#### Select relevant component teams: 
@codice/data 
@codice/ogc 

#### Ask 2 committers to review/merge the PR and tag them here.
@jrnorth @millerw8 

#### How should this be tested?
This condition is covered by the newly added unit tests.  So a successful CI should be sufficient.

#### What are the relevant tickets?
Fixes: #6618

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Threat Dragon models
- [X] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

**Review by kcwire — APPROVED** [review]

**Comment by jlcsmith** [comment]
build now

**Review by jrnorth — APPROVED** [review]

File: catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/GetRecordsMessageBodyReader.java [hunk 1]
```diff
@@ -138,14 +151,16 @@ public CswRecordCollection readFrom(
...
     try {
-      HierarchicalStreamReader reader =
-          new XppReader(
-              new InputStreamReader(inStream, StandardCharsets.UTF_8),
-              XmlPullParserFactory.newInstance().newPullParser());
+      XMLInputFactory xmlInputFactory = XMLInputFactory.newFactory();
+      xmlInputFactory.setProperty(XMLInputFactory.IS_SUPPORTING_EXTERNAL_ENTITIES, false);
+      xmlInputFactory.setProperty(XMLInputFactory.SUPPORT_DTD, false);
+      xmlInputFactory.setProperty(XMLInputFactory.IS_REPLACING_ENTITY_REFERENCES, false);
+      XMLStreamReader xmlStreamReader =
```
[the remark this answers is not in the document]
Comment (jlcsmith) [reply catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/GetRecordsMessageBodyReader.java]: looks like we should be closing both the inputStreamReader and xmlStreamReader

File: catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/GetRecordsMessageBodyReader.java [hunk 2]
```diff
@@ -163,6 +178,11 @@ public CswRecordCollection readFrom(
       throw new WebApplicationException(e, response);
     } finally {
       IOUtils.closeQuietly(inStream);
+      try {
+        xmlStreamReader.close();
```
Comment (jrnorth) [inline catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/GetRecordsMessageBodyReader.java:182]: Should check for null before calling close()

File: catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/GetRecordsMessageBodyReader.java [hunk 3]
```diff
@@ -64,16 +67,26 @@
 
   private DataHolder argumentHolder;
 
+  private QNameMap qNameMap;
+
   public GetRecordsMessageBodyReader(Converter converter, CswSourceConfiguration configuration) {
-    xstream = new XStream(new XppDriver());
+    QName cswQName = new QName(CswConstants.CSW_NAMESPACE_URI, CswConstants.GET_RECORDS_RESPONSE);
```
Comment (millerw8) [inline catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/GetRecordsMessageBodyReader.java:73]: looking at the test records, seems like it all works with test records having both... but how does this handle 
http://www.opengis.net/cat/csw/2.0.2 vs http://www.opengis.net/cat/csw when only 2.0.2 is registered?

Comment (jlcsmith) [reply catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/GetRecordsMessageBodyReader.java:73]: @millerw8  - there is a unit test that covers the CSW 2.0.1 namespace case - https://github.com/codice/ddf/blob/2.19.x/catalog/spatial/csw/spatial-csw-source-common/src/test/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/GetRecordsMessageBodyReaderTest.java#L152.  This one uses the "http://www.opengis.net/cat/csw" namespace.

**Comment by jlcsmith** [comment]
build now

**Comment by jlcsmith** [comment]
build now

File: catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/GetRecordsMessageBodyReader.java [hunk 4]
```diff
@@ -142,10 +147,15 @@ public CswRecordCollection readFrom(
           new XppReader(
               new InputStreamReader(inStream, StandardCharsets.UTF_8),
               XmlPullParserFactory.newInstance().newPullParser());
+
       cswRecords = (CswRecordCollection) xstream.unmarshal(reader, null, argumentHolder);
+    } catch (CannotResolveClassException cre) {
+      // This exception can happen if the XppReader encounters unexpected (but correct) namespaces
+      // prefixes
+      // In this case, attempt to use the StaxReader which is namespace aware.
+      cswRecords = unmarshalWithStaxReader(originalInputStream);
```
Comment (millerw8) [inline catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/GetRecordsMessageBodyReader.java:156]: I am guessing there is a reason why we wouldn't just do it this way all the time?  Is this basically the old way for when it works (registry) and then deal with it here when it doesn't?  I like the less invasive approach, considering our circumstances, just wanted to make sure I understand it.

**Review by millerw8 — APPROVED** [review]

File: catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/GetRecordsMessageBodyReader.java (hunk 4 above)
Comment (jrnorth) [reply catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/GetRecordsMessageBodyReader.java:156]: Yeah, it feels like if there is a way that handles namespaces correctly, then we should use that all the time.

Comment (jlcsmith) [reply catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/reader/GetRecordsMessageBodyReader.java:156]: The StaxReader processes namespace prefix mappings correctly during the unmarshalling, however it doesn't pass the namespace declarations/prefixes to the underlying converter.  This addresses the primary issue we have seen where certain CSW responses use alternative namespace prefix mappings. The CSW Record Converter doesn't care about the namespace prefixes being there (it parses the fields/attributes using the xstream API).  However, the RegistryTransformer utilizes JAXB and relies on well-formed XML including namesapces declarations.  I haven't found a way to tell the StaxReader to preserve the namespace information.   The XppReader does NOT process namespace declarations correctly, but it DOES pass the full XML document to the underlying Converters.  I haven't found one xstream XML reader that works in both cases.

**Comment by jlcsmith** [comment]
Full build passed.  Hero'd with @kcwire by using a CSW server returning non "csw" namespace prefixes.

**Pull request merged**